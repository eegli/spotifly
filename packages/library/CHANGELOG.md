# @spotifly/library

## 0.0.4

### Patch Changes

- 249e235: Fix bin.js in package.json

## 0.0.3

### Patch Changes

- 680b7c4: Minor improvements
- Updated dependencies [680b7c4]
  - @spotifly/utils@0.0.7

## 0.0.2

### Patch Changes

- 9004246: Improve CLIs of all packages and remove unused utils
- Updated dependencies [9004246]
  - @spotifly/utils@0.0.6

## 0.0.1

### Patch Changes

- 4a21f70: Add library package
- Updated dependencies [4a21f70]
  - @spotifly/utils@0.0.5
