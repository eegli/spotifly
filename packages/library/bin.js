#!/usr/bin/env node

require('./cli/dist/spotifly-library-cli.cjs').cli();
