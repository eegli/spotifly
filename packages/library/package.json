{
  "name": "@spotifly/library",
  "version": "0.1.3",
  "description": "Get your Spotify library via CLI",
  "keywords": [
    "spotify",
    "library",
    "music",
    "json",
    "cli",
    "typescript"
  ],
  "homepage": "https://spotifly.nougat.dev/docs/packages/library",
  "repository": {
    "type": "git",
    "url": "https://github.com/eegli/spotifly"
  },
  "license": "MIT",
  "author": {
    "name": "Eric Egli",
    "email": "lx180777+npm@gmail.com"
  },
  "main": "dist/spotifly-library.cjs.js",
  "module": "dist/spotifly-library.esm.js",
  "files": [
    "dist",
    "cli"
  ],
  "dependencies": {
    "@eegli/tinyparse": "^0.6.0",
    "@spotifly/core": "workspace:^",
    "@spotifly/utils": "workspace:^",
    "cli-progress": "^3.11.2"
  },
  "devDependencies": {
    "@types/cli-progress": "^3.11.0"
  },
  "engines": {
    "node": ">=14"
  },
  "publishConfig": {
    "directory": "_release/package"
  },
  "preconstruct": {
    "entrypoints": [
      "index.ts",
      "cli.ts"
    ]
  }
}
