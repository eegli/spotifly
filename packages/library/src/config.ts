import { createParser } from '@eegli/tinyparse';
import { DefaultConfig } from './types';

export const defaultConfig: DefaultConfig = {
  token: '',
  type: 'light',
  genres: false,
  features: false,
  compact: false,
  outDir: '',
};

export const { parse, help } = createParser(defaultConfig, [
  {
    name: 'token',
    required: true,
    description:
      'A Spotify user access token. Requires at least the scope "user-library-read"',
  },
  {
    name: 'type',
    description:
      "Output type per track. Either 'full' or 'light'. Default: 'light'",
  },
  {
    name: 'genres',
    description: 'Include artist genres for each track. Default: false',
  },
  {
    name: 'features',
    description: 'Include audio features for each track. Default: false',
  },
  {
    name: 'compact',
    description:
      'Output more compact/minified JSON and save disk space. Default: false',
  },
  {
    name: 'outDir',
    description: 'Custom relative output directory. Default: Current directory',
  },
]);
