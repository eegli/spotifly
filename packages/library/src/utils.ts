import cliProgress from 'cli-progress';

export const createProgressBar = (items: string) =>
  new cliProgress.SingleBar({
    format:
      `Fetching ${items}\t |` +
      '{bar} {percentage}% | ETA: {eta}s | {value}/{total}',
    barCompleteChar: '\u25A0',
    barIncompleteChar: ' ',
    hideCursor: true,
  });

/**
 * Checks if the first date as a string is before another date
 */
export const isBeforeDate = (
  date1: string | Date,
  date2: string | Date,
): boolean => {
  const d1 = date1 instanceof Date ? date1.getTime() : Date.parse(date1);
  const d2 = date2 instanceof Date ? date2.getTime() : Date.parse(date2);
  if (isNaN(d1) || isNaN(d2)) {
    return false;
  }
  return d1 < d2;
};
