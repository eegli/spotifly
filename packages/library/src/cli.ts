import { ValidationError } from '@eegli/tinyparse';
import { colors } from '@spotifly/utils';
import pkg from '../package.json';
import { help, parse } from './config';
import { getLibrary } from './library';

export const execute = async (): Promise<void> => {
  try {
    const config = await parse(process.argv.slice(2));
    await getLibrary(config);
  } catch (err) {
    if (err instanceof ValidationError) {
      console.error(colors.red('Error: ' + err.message));
    } else {
      throw err;
    }
  }
};

export const cli = async (): Promise<void> => {
  if (process.argv.length < 3) {
    console.info(
      `
${colors.yellow(`${pkg.name} v${pkg.version}`)}

${help('CLI arguments')}

For docs & help, visit ${pkg.homepage}.
`
    );
    return;
  }
  return execute();
};
