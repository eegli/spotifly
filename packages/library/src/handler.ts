import { initialize, isError as isSpotiflyError } from '@spotifly/core';
import { writeJSON } from '@spotifly/utils/fs';
import log from '@spotifly/utils/log';
import { Result } from '@spotifly/utils/types';
import type { Library, LibraryExport, LibraryParams } from './types';
import { createProgressBar, isBeforeDate } from './utils';

export const libraryHandler = async ({
  options,
  globals,
}: LibraryParams): Promise<Result<LibraryExport>> => {
  let progress = createProgressBar('user library');
  try {
    const spotifyClient = initialize({ accessToken: globals.token });

    let library: Library = [];

    progress.start(0, 0);

    let nextPage: string | null = null;
    let offset = 0;

    fetchLoop: do {
      const { data } = await spotifyClient.Tracks.getUsersSavedTracks({
        limit: 50,
        offset,
      });
      progress.setTotal(data.total);
      progress.increment(data.items.length);

      for (let i = 0; i < data.items.length; i++) {
        const track = data.items[i];
        if (library.length === options.last) {
          break fetchLoop;
        }
        if (isBeforeDate(track.added_at, options.since)) {
          break fetchLoop;
        }
        library.push(track);
      }
      nextPage = data.next;
      offset += 50;
    } while (nextPage);

    progress.stop();

    // Reduce library if necessary
    if (options.type === 'light') {
      library = library.reduce(
        (acc, curr) => {
          acc.push({
            added_at: curr.added_at,
            track: {
              id: curr.track.id,
              name: curr.track.name,
              album: {
                id: curr.track.album.id,
                name: curr.track.album.name,
              },
              artists: curr.track.artists.map(a => ({
                id: a.id,
                name: a.name,
              })),
            },
          });
          return acc;
        },
        <Library>[],
      );
    }

    // Add genres if specified
    if (options.genres) {
      const artists = library.map(t => t.track.artists.map(a => a.id)).flat();
      const artistIds = [...new Set<string>(artists)];
      const genres: Record<string, string[]> = {};

      progress = createProgressBar('artists');
      progress.start(artistIds.length, 0);

      await spotifyClient.Artists.getAllArtists(artistIds)(({ data }) => {
        data.artists.forEach(artist => {
          genres[artist.id] = artist.genres;
        });
        progress.increment(data.artists.length);
      });

      progress.stop();
      library.forEach(({ track }) => {
        track.genres = track.artists.map(({ id }) => genres[id]);
      });
    }

    // Add audio features if specified
    if (options.features) {
      const trackIds = library.map(t => t.track.id);
      const features: Record<string, SpotifyApi.AudioFeaturesObject> = {};

      progress = createProgressBar('audio features');

      progress.start(trackIds.length, 0);

      await spotifyClient.Tracks.getAllAudioFeatures(trackIds)(({ data }) => {
        data.audio_features.forEach(f => {
          features[f.id] = f;
        });
        progress.increment(data.audio_features.length);
      });

      progress.stop();
      library.forEach(({ track }) => {
        track.features = features[track.id];
      });
    }

    const libExport: LibraryExport = {
      meta: {
        date_generated: new Date().toISOString(),
      },
      library,
    };
    const fileName = 'spotify-library';
    const outFile = await writeJSON({
      fileName,
      path: options.outDir,
      data: libExport,
      compact: options.compact,
    });

    log.info(`Success! Library written to ${outFile}`);
    return {
      success: true,
      value: libExport,
    };
  } catch (error) {
    progress.stop();
    if (isSpotiflyError(error) && error.response) {
      const { status, message } = error.response.data.error;
      return {
        success: false,
        error: `Failed to call Spotify API: Status ${status}, ${message}`,
      };
    } else if (error instanceof Error) {
      return {
        success: false,
        error: error.message,
      };
    } else {
      return {
        success: false,
        error: 'Something went wrong',
      };
    }
  }
};
