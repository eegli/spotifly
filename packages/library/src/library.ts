import { writeJSON } from '@spotifly/utils';
import axios from 'axios';
import { defaultConfig } from './config';
import type {
  Library,
  LibraryExport,
  MultipleArtists,
  MultipleAudioFeatures,
  Options,
  TrackFull,
  TrackLight,
  UsersSavedTracks,
} from './types';
import { chunkify, createProgressBar } from './utils';

export const getLibrary = async (
  options: Options
): Promise<LibraryExport<TrackLight | TrackFull>> => {
  try {
    const config = { ...defaultConfig, ...options };
    const lib: LibraryExport<TrackFull | TrackLight> = {
      meta: {
        date_generated: new Date().toISOString(),
        output_type: config.type,
      },
      library: [],
    };
    let progress = createProgressBar('user library');

    const pagination = {
      nextUrl: <string | null>'https://api.spotify.com/v1/me/tracks',
      start: 0,
      end: 50,
    };

    progress.start(0, 0);
    while (pagination.nextUrl) {
      const { data }: UsersSavedTracks = await axios.get(pagination.nextUrl, {
        headers: {
          Authorization: `Bearer ${config.token}`,
        },
        params: {
          limit: 50,
        },
      });

      progress.setTotal(data.total);
      progress.increment(data.items.length);

      data.items.forEach(el => {
        lib.library.push(el);
      });

      pagination.nextUrl = data.next;
      pagination.start = pagination.end;
      pagination.end += data.limit;
    }

    progress.stop();

    // Reduce library if necessary
    if (config.type === 'light') {
      lib.library = lib.library.reduce((acc, curr) => {
        acc.push({
          added_at: curr.added_at,
          track: {
            id: curr.track.id,
            name: curr.track.name,
            album: {
              id: curr.track.album.id,
              name: curr.track.album.name,
            },
            artists: curr.track.artists.map(a => ({
              id: a.id,
              name: a.name,
            })),
          },
        });
        return acc;
      }, <Library<TrackLight>>[]);
    }

    // Add genres if specified
    if (config.genres) {
      const artists = lib.library
        .map(t => t.track.artists.map(a => a.id))
        .flat();
      const artistIds = new Set<string>(artists);
      const genres: Record<string, string[]> = {};

      // Max 50 items per request to Spotify
      const chunkSize = 50;

      progress = createProgressBar('artists');
      progress.start(artistIds.size, 0);

      for await (const chunk of chunkify(artistIds, chunkSize)) {
        const { data }: MultipleArtists = await axios.get(
          'https://api.spotify.com/v1/artists',
          {
            headers: {
              Authorization: `Bearer ${config.token}`,
            },
            params: {
              ids: chunk.join(','),
            },
          }
        );
        data.artists.forEach(artist => {
          genres[artist.id] = artist.genres;
        });
        progress.increment(chunkSize);
      }
      progress.stop();
      lib.library.forEach(({ track }) => {
        track.genres = track.artists.map(({ id }) => genres[id]);
      });
    }

    // Add audio features if specified
    if (config.features) {
      const songIds = lib.library.map(t => t.track.id);
      const features: Record<string, SpotifyApi.AudioFeaturesObject> = {};

      const chunkSize = 50;
      progress = createProgressBar('audio features');

      progress.start(songIds.length, 0);
      for await (const chunk of chunkify(songIds, 50)) {
        const { data }: MultipleAudioFeatures = await axios.get(
          'https://api.spotify.com/v1/audio-features',
          {
            headers: {
              Authorization: `Bearer ${config.token}`,
            },
            params: {
              ids: chunk.join(','),
            },
          }
        );

        data.audio_features.forEach(f => {
          features[f.id] = f;
        });
        progress.increment(chunkSize);
      }
      progress.stop();
      lib.library.forEach(({ track }) => {
        track.features = features[track.id];
      });
    }

    const outDir = await writeJSON({
      fileName: 'spotify-library',
      path: config.outDir,
      data: lib,
      compact: config.compact,
    });
    console.info("Success! Library written to '%s'", outDir);
    return lib;
  } catch (e) {
    if (axios.isAxiosError(e)) {
      const error = new Error(e.message);
      Object.assign(error, e.toJSON());
      throw error;
    }
    throw e;
  }
};
