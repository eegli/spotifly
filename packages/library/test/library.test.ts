import * as utils from '@spotifly/utils';
import axios from 'axios';
import { getLibrary } from '../src/library';
import {
  RES_MULTIPLE_ARTISTS,
  RES_MULTIPLE_AUDIO_FEATURES,
  RES_USER_SAVED_TRACKS,
} from './fixtures';

jest.mock('axios');

const writeSpy = jest
  .spyOn(utils, 'writeJSON')
  .mockImplementation(opts => Promise.resolve(opts.path));
const mockAxios = axios as jest.Mocked<typeof axios>;

beforeEach(() => {
  jest.clearAllMocks();
});

describe('Library', () => {
  it('gets light library', async () => {
    mockAxios.get.mockResolvedValueOnce(RES_USER_SAVED_TRACKS);
    const res = await getLibrary({ token: '123x' });
    expect(res.meta.output_type).toBe('light');
    expect(res.meta.date_generated).toEqual(expect.any(String));
    expect(res.library[0]).toMatchSnapshot('light track');
    expect(mockAxios.get).toHaveBeenCalledTimes(1);
  });
  it('gets full library', async () => {
    mockAxios.get.mockResolvedValueOnce(RES_USER_SAVED_TRACKS);
    const res = await getLibrary({
      token: '123x',
      type: 'full',
    });
    expect(res.meta.output_type).toBe('full');
    expect(res.meta.date_generated).toEqual(expect.any(String));
    expect(res.library[0]).toMatchObject(RES_USER_SAVED_TRACKS.data.items[0]);
    expect(mockAxios.get).toHaveBeenCalledTimes(1);
  });
  it('with genres', async () => {
    mockAxios.get
      .mockResolvedValueOnce(RES_USER_SAVED_TRACKS)
      .mockResolvedValueOnce(RES_MULTIPLE_ARTISTS);
    const res = await getLibrary({
      token: '123x',
      genres: true,
    });
    expect(res.library[0].track.genres).toBeTruthy();
    expect(res.library[0].track.genres).toMatchSnapshot('genres');
    expect(mockAxios.get).toHaveBeenCalledTimes(2);
  });
  it('with audio features', async () => {
    mockAxios.get
      .mockResolvedValueOnce(RES_USER_SAVED_TRACKS)
      .mockResolvedValueOnce(RES_MULTIPLE_AUDIO_FEATURES);
    const res = await getLibrary({
      token: '123x',
      features: true,
    });
    expect(res.library[0].track.features).toBeTruthy();
    expect(res.library[0].track.features).toMatchSnapshot('features');
    expect(mockAxios.get).toHaveBeenCalledTimes(2);
  });
  it('writes library', async () => {
    mockAxios.get.mockResolvedValueOnce(RES_USER_SAVED_TRACKS);
    await getLibrary({
      token: '123x',
    });
    expect(writeSpy).toHaveBeenCalledTimes(1);
  });
});
