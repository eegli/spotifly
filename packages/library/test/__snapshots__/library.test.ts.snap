// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Library gets light library: light track 1`] = `
Object {
  "added_at": "2021-09-30T22:04:57Z",
  "track": Object {
    "album": Object {
      "id": "0vlWyzbAl3ff9b3zout1XQ",
      "name": "Stay With Me",
    },
    "artists": Array [
      Object {
        "id": "5Ho1vKl1Uz8bJlk4vbmvmf",
        "name": "Amelie Lens",
      },
    ],
    "id": "5UsfWcP6SThHlZ4oAgx7ge",
    "name": "Follow",
  },
}
`;

exports[`Library with audio features: features 1`] = `
Object {
  "acousticness": 0.0375,
  "analysis_url": "https://api.spotify.com/v1/audio-analysis/5UsfWcP6SThHlZ4oAgx7ge",
  "danceability": 0.725,
  "duration_ms": 337500,
  "energy": 0.908,
  "id": "5UsfWcP6SThHlZ4oAgx7ge",
  "instrumentalness": 0.822,
  "key": 7,
  "liveness": 0.108,
  "loudness": -8.798,
  "mode": 1,
  "speechiness": 0.0345,
  "tempo": 128.006,
  "time_signature": 4,
  "track_href": "https://api.spotify.com/v1/tracks/5UsfWcP6SThHlZ4oAgx7ge",
  "type": "audio_features",
  "uri": "spotify:track:5UsfWcP6SThHlZ4oAgx7ge",
  "valence": 0.134,
}
`;

exports[`Library with genres: genres 1`] = `
Array [
  Array [
    "acid techno",
    "belgian techno",
    "dark techno",
    "electra",
  ],
]
`;
